### ISSUE TITLE
Bad error message for DistributedSampler when num_replicas is too high

### ISSUE THREAD

**Issue by alanhdu** [pytorch/pytorch#49172]
## 🐛 Bug

When using `DistributedSampler` with a very small dataset, you get an assertion error:

## To Reproduce

```python
import torch

class RandomDataset(torch.utils.data.Dataset):
    def __init__(self, length: int):
        self.length = length
    def __len__(self):
        return self.length
    def __getitem__(self, idx: int):
        return torch.rand(100), torch.rand(5)

num_replicas = 7    # 6 works fine
for i in range(num_replicas):
    sampler = torch.utils.data.DistributedSampler(
        RandomDataset(3), num_replicas=num_replicas, rank=i
    )
    print(list(sampler))
```

returns 

```
Traceback (most recent call last):
  File "scratch.py", line 16, in <module>
    print(list(sampler))
  File "/opt/anaconda/envs/ctrldev/lib/python3.6/site-packages/torch/utils/data/distributed.py", line 107, in __iter__
    assert len(indices) == self.total_size
AssertionError
```

## Expected behavior

Either the sample duplication behavior should work here too, or we should raise an error in the constructor.

## Environment

```
Collecting environment information...
PyTorch version: 1.7.0
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Fedora release 32 (Thirty Two) (x86_64)
GCC version: (GCC) 10.2.1 20201016 (Red Hat 10.2.1-6)
Clang version: 10.0.1 (Fedora 10.0.1-3.fc32)
CMake version: Could not collect

Python version: 3.6 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.16.5
[pip3] numpy-quaternion==2020.9.5.14.42.2
[pip3] pytorch-lightning==1.0.8
[pip3] torch==1.7.0
[pip3] torchsummary==1.5.1
[conda] blas                      2.20                        mkl    conda-forge
[conda] cpuonly                   1.0                           0    pytorch
[conda] libblas                   3.8.0                    20_mkl    conda-forge
[conda] libcblas                  3.8.0                    20_mkl    conda-forge
[conda] liblapack                 3.8.0                    20_mkl    conda-forge
[conda] liblapacke                3.8.0                    20_mkl    conda-forge
[conda] mkl                       2020.4             h726a3e6_304    conda-forge
[conda] numpy                     1.16.5           py36h95a1406_0    conda-forge
[conda] pytorch                   1.7.0               py3.6_cpu_0  [cpuonly]  pytorch
[conda] pytorch-lightning         1.0.8              pyhd8ed1ab_0    conda-forge
[conda] torchsummary              1.5.1                    pypi_0    pypi
```

cc @ezyang @gchanan @zou3519 @bdhirsh @SsnL @VitalyFedyunin @ejguan @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd

**Comment by alanhdu** [comment]
Related: now that we have https://github.com/pytorch/pytorch/issues/38174, could we add a flag to `DistributedDataSampler` that just allows uneven partitions across the replicas?

**Comment by agolynski** [comment]
cc @VitalyFedyunin

**Comment by ejguan** [comment]
I think it's already been solved by #45329
https://github.com/pytorch/pytorch/blob/f4e15c4a234d1d4ccba0ff8c7f53a59a36e88edb/torch/utils/data/distributed.py#L107-L111
Maybe try to use nightly or build from source.

**Comment by peterbell10** [comment]
Can confirm: reproducer fails on 1.7 but passes on a recent master.
